src/conversations/Conversations.ts

```diff
@@ -97,56 +97,115 @@ export default class Conversations {
+
+      // Decode as an invite and return the envelope
+      // This gives the contentTopicUpdater everything it needs to add to the topic list
+      if (contentTopic === inviteTopic) {
+        const sealed = await SealedInvitation.fromEnvelope(env)
+        const unsealed = await sealed.v1.getInvitation(this.client.keys)
+        return ConversationV2.create(this.client, unsealed, sealed.v1.header)
+      }
+
+      const convo = convoMap.get(contentTopic)
+
+      // Decode as a V1 message if the topic matches a V1 convo
+      if (convo instanceof ConversationV1) {
+        return this.client.decodeEnvelope(env)
       }
-      dmAddresses.push(peerAddress)
-      return this.buildTopicsForAllMessages(dmAddresses, introTopic)
+
+      // Decode as a V2 message if the topic matches a V2 convo
+      if (convo instanceof ConversationV2) {
+        return convo.decodeMessage(env)
+      }
+
+      console.log('Unknown topic')
+
+      throw new Error('Unknown topic')
     }
 
-    // Filter intro topics if already streaming direct messages for that address to avoid duplicates
-    const filter = (msg: MessageV1): boolean => {
-      if (
-        msg.contentTopic === introTopic &&
-        messageHasHeaders(msg) &&
-        dmAddresses.includes(this.getPeerAddress(msg))
-      ) {
-        return false
+    const contentTopicUpdater = (msg: Conversation | Message | null) => {
+      // If we have a V1 message from the introTopic, store the conversation in our mapping
+      if (msg instanceof MessageV1 && msg.contentTopic === introTopic) {
+        const convo = new ConversationV1(
+          this.client,
+          msg.recipientAddress === this.client.address
+            ? (msg.senderAddress as string)
+            : (msg.recipientAddress as string),
+          msg.sent
+        )
+
+        if (topics.has(convo.topic)) {
+          return undefined
+        }
+
+        convoMap.set(convo.topic, convo)
+        topics.add(convo.topic)
+
+        return Array.from(topics.values())
       }
-      return true
+
+      if (msg instanceof ConversationV2) {
+        if (topics.has(msg.topic)) {
+          return undefined
+        }
+
+        convoMap.set(msg.topic, msg)
+        topics.add(msg.topic)
+
+        return Array.from(topics.values())
+      }
+
+      return undefined
     }
 
-    return Stream.create<MessageV1>(
+    const str = await Stream.create<Message | Conversation | null>(
       this.client,
-      topics,
-      async (env) => {
-        const msg = await this.client.decodeEnvelope(env)
-        return filter(msg) ? msg : undefined
-      },
+      Array.from(topics.values()),
+      decodeMessage,
       contentTopicUpdater
     )
+
+    return (async function* generate() {
+      for await (const val of str) {
+        if (val instanceof MessageV1 || val instanceof MessageV2) {
+          yield val
+        }
+        // For conversation V2, we know we have already missed the message when we get the invite
```

Comment:
I think this statement is a bit oversimplified. With V2 the invite is actually sent earlier than with V1, we send invites from `newConversation` (V1 does nothing here), the first message is sent with `send` after that. That of course doesn't mean that we didn't miss the message by the time we subscribe to the topic. Same could easily be true with V1 convos, by the time we subscribe in response to the first message, the second, third, etc message could have already gone through.

I'm not quite sure what sort of guarantees is reasonable to offer here. But it's definitely more complicated than what we admit here.

Reply:
That's fair. We don't know for sure if we have missed anything, but we do know that if there is anything in the topic that we will likely not have captured it at this point.

I can clarify the comment